File: R/tools/math.R
import::here(tidyr, 'pivot_wider')
import::here(file.path(wd, 'R', 'tools', 'df_tools.R'),
    'pivot_then_collapse',
    .character_only=TRUE)
import::here(file.path(wd, 'R', 'tools', 'list_tools.R'),
    'collect_matrix_cols', 'matrix2list', 'fill_missing_keys',
    .character_only=TRUE)

## Functions
## unpaired_t_test
## apply_unpaired_t_test
## fishers_lsd
## tukey_multiple_comparisons
## bonferroni_multiple_comparisons
## apply_multiple_comparisons
## generate_normal_data
## generate_lognormal_data
## compute_normal_tvd
## compute_lognormal_tvd


#' Unpaired T Test
#' 
#' @description Use this when you have multiple groups that are not dependent on each other
#' and the variances are clearly unequal between the groups
#' Try this as a second resort if fishers_lsd is giving too many false positives
#'
unpaired_t_test <- function(x, y) {

    # exit if either side is null or both sides have fewer observations
    if (is.null(x) | is.null(y)) {
        return(NA)
    }
    if ((length(x)==1) & (length(y)==1)) {
        return(NA)
    }

    # two.sided if x and y both have multiple observations
    if ((length(x)>1) & (length(y)>1)) {
        pval <- t.test(x, y, alternative='two.sided', var.equal=FALSE)[['p.value']]
        return(pval)
    }

    # if one side only has a single observation, use one.sided t test 
    if (length(x)==1) {
        ob <- x  # single observation
        obs <- y
    } else if (length(y)==1) {
        obs <- x 
        ob <- y  # single observation
    }
    pval <- t.test(
        obs,
        alternative=(if (mean(obs) > ob) {"greater"} else {"less"}),
        mu=ob,
        var.equal=FALSE
    )[['p.value']]

    return(pval)
}


#' Apply Unpaired T Test
#' 
#' @description Apply unpaired t test to dataframe
#' 
apply_unpaired_t_test <- function(
    df,
    index_cols,  # c('organ', 'cell_type')
    group_name,  # 'group_name'
    metric,  # 'pct_cells'
    custom_group_order=c()
) {

    if (all(is.na( df[[metric]] )) ) {
        return(NA)
    }

    if (length(custom_group_order)>=1) {
        group_names <- intersect( custom_group_order, unique(df[[group_name]]) )
    } else {
        group_names <- sort(unique( df[[group_name]] ))
    }

    n_groups <- length(group_names)

    # exit if only one group
    if (n_groups==1) {
        return(NaN)
    }
    
    # mfi_tbl <- pivot_then_collapse(
    #     df,
    #     index_cols=c('organ', 'cell_type'),
    #     group_name='group_name',
    #     metric=opt[['metric']]
    # )
    res <- pivot_then_collapse(
        df, index_cols, group_name,
        metric, custom_group_order
    )

    # iterate through all pairs of columns
    # collect results directly in pval_col
    id_combos <- collect_matrix_cols(combn(1:n_groups, 2))  # generate pairs of indexes
    pval_cols <- sapply(id_combos,  
        function(x) paste(group_names[x[[2]]], group_names[x[[1]]], sep='-')
    )  # generate name pairs
    for (idx in 1:choose(n_groups, 2)) {
        idx1 <- id_combos[[idx]][1]  # 1st col
        idx2 <- id_combos[[idx]][2]  # 2nd col

        # t test
        res[ pval_cols[idx] ] <- mapply(
            function(x, y) unpaired_t_test(x, y),
            res[[ group_names[idx1] ]],  # 1st col
            res[[ group_names[idx2] ]]   # 2nd col
        )
    }

    return(res)
}


#' One Way ANOVA with no correction, ie. Fisher's LSD
#' 
#' @description Use this when you have multiple groups that are not dependent on each other
#' and the variances are roughly equal between the groups.
#' Use this first. If the group with the largest variance is more than 4x the variance of
#' the smallest group, switch to unpaired_t_test
#' 
fishers_lsd <- function(
    df,
    group,  # 'group_name'
    metric,  # 'pct_cells'
    custom_group_order=c()
) {

    if (all(is.na( df[[metric]] )) ) {
        return(NA)
    }

    if (length(custom_group_order)>=1) {
        group_names <- intersect( custom_group_order, unique(df[[group]]) )
    } else {
        group_names <- sort(unique( df[[group]] ))
    }

    n_groups <- length(group_names)

    # exit if only one group
    if (n_groups==1) {
        return(NA)
    }


    # ----------------------------------------------------------------------
    # ANOVA

    # computations
    formula <- as.formula(
        paste(deparse(as.name(metric)), "~", deparse(as.name(group)))  # 'metric ~ group'
    )
    group_means <- aggregate(formula, data = df, FUN = mean)
    group_means <- setNames(group_means[[metric]], group_means[[group]])
    group_sizes <- aggregate(formula, data = df, FUN = length)
    group_sizes <- setNames(group_sizes[[metric]], group_sizes[[group]])

    overall_mean <- mean(df[[metric]])
    ss_total <- sum((df[[metric]] - overall_mean)^2)
    ss_between <- sum(group_sizes * (group_means - overall_mean)^2)
    ss_within <- ss_total - ss_between

    df_between <- length(group_names) - 1  # degrees of freedom between
    df_within <- length(df[[metric]]) - length(group_names)  # degrees of freedom within
    mse <- ss_within / df_within  # mean square error


    # ----------------------------------------------------------------------
    # Fisher's LSD p values

    # iterate through all pairs of columns
    # collect results directly in pval_col
    id_combos <- collect_matrix_cols(combn(1:n_groups, 2))  # generate pairs of indexes
    pval_cols <- sapply(id_combos,  
        function(x) paste(group_names[x[[2]]], group_names[x[[1]]], sep='-')
    )  # generate name pairs
    pvals <- list()

    for (idx in 1:choose(n_groups, 2)) {
        left <- group_names[ id_combos[[idx]][1] ]  # 1st col
        right <- group_names[ id_combos[[idx]][2] ]  # 2st col

        # fishers lsd test
        mean_diff <- group_means[right] - group_means[left]
        se_diff <- sqrt(mse * (1/group_sizes[left] + 1/group_sizes[right]))
        t_stat <- mean_diff / se_diff
        pval <- 2 * pt(-abs(t_stat), df_within)

        pvals[ pval_cols[idx] ] <- pval
    }

    return(pvals)
}


#' One Way ANOVA with Tukey correction
#' 
#' @description Use this when you have multiple groups that are may be dependent on each other
#' and the variances are roughly equal between the groups. This is supposed to control for false
#' positives due to group effects. In general, this will provide larger p values
#' (ie. less significance) compared with fishers_lsd.
#' 
tukey_multiple_comparisons <- function(
    df,
    group,  # 'group_name',
    metric,  # 'pct_cells'
    custom_group_order=c()
) {

    if (all(is.na( df[[metric]] )) ) {
        return(NA)
    }

    if (length(custom_group_order)>=1) {
        group_names <- intersect( custom_group_order, unique(df[[group]]) )
        df[[group]] <- factor(df[[group]], levels = custom_group_order)
    } else {
        group_names <- sort(unique( df[[group]] ))
    }

    n_groups <- length(group_names)

    # exit if only one group
    if (n_groups==1) {
        return(NA)
    }

    formula <- as.formula(
        paste(deparse(as.name(metric)), "~", deparse(as.name(group)))  # 'metric ~ group'
    )
    withCallingHandlers({
        fit <- aov(formula, data = df)
        pvals <- TukeyHSD(fit)[[group]][, 'p adj']
    }, warning = function(w) {
        if ( any(grepl("NaNs produced", w)) ) {
            invokeRestart("muffleWarning")
        }
    })

    if (n_groups==2) {
        pvals <- setNames(pvals, paste(group_names[2],group_names[1], sep='-'))
    }

    return(pvals)
}


#' One Way ANOVA with Bonferroni correction
#' 
#' @description Use this when you have multiple groups that are heavily dependent on each other
#' and there are a large number of false positives because of group interactions. Since this is
#' much more stringent than tukey_multiple_comparisons, this is not recommended due to the
#' potential to have false negatives, especially for small datasets.
#' 
bonferroni_multiple_comparisons <- function(
    df,
    group,  # 'group_name'
    metric,  # 'pct_cells'
    custom_group_order=c()
) {

    if (all(is.na( df[[metric]] )) ) {
        return(NA)
    }

    if (length(custom_group_order)>=1) {
        group_names <- intersect( custom_group_order, unique(df[[group]]) )
        df[[group]] <- factor(df[[group]], levels = custom_group_order)
    } else {
        group_names <- sort(unique( df[[group]] ))
    }

    n_groups <- length(unique( df[[group]] ))

    # exit if only one group
    if (n_groups==1) {
        return(NA)
    }

    pvals_mat <- pairwise.t.test(
        df[[metric]], df[[group]],
        p.adjust.method = "bonferroni"
    )[['p.value']]
    pvals <- matrix2list(pvals_mat)  # flatten
    # pvals <- Filter(Negate(is.na), pvals)  # filter NA

    return(pvals)
}


#' Apply Multiple Comparisons
#' 
#' @description Applies one-way ANOVA across a dataframe
#' Currently, this implementation skips the ANOVA parameters and instead just
#' provides the p_values of the post-hoc analysis for plotting
#' In the future, the ANOVA F-statistic may be provided
#' 
apply_multiple_comparisons <- function(
    df,
    index_cols,  # c('organ', 'cell_type')
    group_name,  # 'group_name'
    metric,  # 'pct_cells' or 'abs_count'
    correction='fishers_lsd',  # 'fishers_lsd', 'tukey', or 'bonferroni'
    custom_group_order=c()
) {

    res <- pivot_then_collapse(
        df, index_cols, group_name,
        metric, custom_group_order
    )

    # split dataframe to fit expected input
    df_list <- split(
        df[, c(index_cols, group_name, metric)],
        df[, c(index_cols)],
        drop=TRUE
    )

    if (correction=='fishers_lsd') {
        pval_list <- mapply(
            function(x) fishers_lsd(x, group=group_name, metric=metric),
            df_list, SIMPLIFY=FALSE
        )
    } else if (correction=='tukey') {
        pval_list <- mapply(
            function(x) tukey_multiple_comparisons(x, group=group_name, metric=metric),
            df_list, SIMPLIFY=FALSE
        )
    } else if (correction=='bonferroni') {
        pval_list <- mapply(
            function(x) bonferroni_multiple_comparisons(x, group=group_name, metric=metric),
            df_list, SIMPLIFY=FALSE
        )
    } else {
        stop("Choose correction='fishers_lsd', 'tukey' or 'bonferroni'")
    }

    # convert to matrix
    colnames <- unique(unlist(lapply(pval_list, names)))  # get all column names
    if (length(colnames)==1) {
        pvals <- matrix(pval_list, nrow=1, dimnames=list(colnames, names(pval_list)))
    } else if (is.null(dim(pval_list))) {
        pvals <- mapply(function(x) fill_missing_keys(x, colnames), pval_list)
    }
    pvals <- matrix(
        sapply(pvals, function(x) x[[1]] ),
        nrow = nrow(pvals), ncol = ncol(pvals), dimnames = dimnames(pvals)
    )

    res <- cbind(res, t(pvals))  # res and pvals are sorted for the same order
    res <- res[do.call(order, res[index_cols]), ]  # sort rows in original order

    return(res)
}


#' Generate Normal Data
#' 
#' Generate a dataframe of random values drawn from a normal distribution
#' Use this as an input to plot_modal_histograms
#' 
generate_normal_data <- function(
    n=1000, mean=200, sd=100,
    group_name="group"
) {
    df <- data.frame(group=group_name, value=rnorm(n, mean=mean, sd=sd))
    return(df)
}


#' Generate Lognormal Data
#' 
#' Use this function to simulate data for building fluorescence histograms
#' 
#' Flow intensity data is lognormal
#' See this blog post on how to correctly use the rlnorm function
#' https://msalganik.wordpress.com/2017/01/21/making-sense-of-the-rlnorm-function-in-r/
#' 
#' Special care had to be used to handle negative MFI values
#' When Flowjo reports a negative MFI, the actual intensity is 1/abs(MFI)
#' To correct for this, I convert the MFI values using the following formula:
#' adj_mean <- 10^(sign(MFI) * log10(1 + abs(MFI)))
#' This formula was chosen so that values in the range -1 < MFI < 1 are smooth and continuous
#' For very large positive values of MFI, the formula approaches adj_mean = MFI
#' For very negative values of MFI, the formula approaches adj_mean <- 1/MFI
#' 
generate_lognormal_data <- function(
  n=1000, mean=200, sd=100, group_name = "group"
) {

    if (mean==0) {
        adj_mean <- 1
    } else {
        # make the mean continuous
        signed_log <- sign(mean) * log10(1 + abs(mean))
        adj_mean <- 10^signed_log
    }
    adj_sd <- abs(sd/mean) * adj_mean

    location <- log(adj_mean^2 / sqrt(adj_sd^2 + adj_mean^2))
    shape <- sqrt(log(1 + (adj_sd / adj_mean)^2))

    # clamp mean to prevent errors
    if (location < -372.5666) {
        location <- -372.5666
    }

    values <- rlnorm(n, meanlog = location, sdlog = shape)
    df <- data.frame(group = group_name, value = values)

    return(df)
}


#' Compute Normal Total Variation Distance
#'
#' Returns a value between 0 and 1 that quantifies 1-overlap. This can be
#' interpreted as dissimilarity between two distributions. To enable this 
#' function handle to negative MFI values, the same log transform used for
#' generate_lognormal_data has been added. This reduces the differences
#' when negative MFI values are compared.
#' 
compute_normal_tvd <- function(mean1, sd1, mean2, sd2, log_transform=FALSE) {

    if (is.vector(mean1) || is.list(mean1)) {
        mean1 <- mean(mean1)
    }
    if (is.vector(sd1) || is.list(sd1)) {
        sd1 <- mean(unlist(sd1))
    }
    if (is.vector(mean2) || is.list(mean2)) {
        mean2 <- mean(unlist(mean2))
    }
    if (is.vector(sd2) || is.list(sd2)) {
        sd2 <- mean(unlist(sd2))
    }

    if (is.na(sd1) & is.na(sd2)) { return(NA) }
    if (is.na(mean2) & is.na(sd2)) { return(NA) }
    if (!is.na(sd1) & is.na(sd2)) { sd2 <- sd1 }
    if (is.na(sd1) & !is.na(sd2)) { sd1 <- sd2 }
    if (sd1==0 | sd2==0) { return(NA) }
    if (is.na(mean2)) { mean2 <- 0 }

    if (log_transform) {
        mean1 <- 10^(sign(mean1)*log10(1+abs(mean1))) /log(10)
        sd1 <- sd1/log(10)
        mean2 <- 10^(sign(mean2)*log10(1+abs(mean2))) /log(10)
        sd2<- sd2/log(10)
    }

    p <- function(x) dnorm(x, mean = mean1, sd = sd1)
    q <- function(x) dnorm(x, mean = mean2, sd = sd2)
    min_pdf <- function(x) pmin(p(x), q(x))  # Select minimum of two curves

    # Get area under the curve of the overlap
    # Set integral limit at 10*sd in each direction
    lower <- min(mean1 - 10 * sd1, mean2 - 10 * sd2)
    upper <- max(mean1 + 10 * sd1, mean2 + 10 * sd2)
    overlap <- integrate(min_pdf, lower, upper)$value

    dissimiliarity <- 1 - overlap
    return(dissimiliarity)
}


#' Compute Lognormal Total Variation Distance
#' 
#' Returns a value between 0 and 1 that quantifies 1-overlap. This can be
#' interpreted as dissimilarity between two distributions. Since this is
#' NOT computed on linear space, this tends to magnify differences when
#' one mean is large and the other is small. Therefore, compute_normal_tvd
#' is a better metric for measuring overlap for fluorescence data.
#' 
compute_lognormal_tvd <- function(mean1, sd1, mean2, sd2) {

    if (is.na(sd1) & is.na(sd2)) { return(NA) }
    if (is.na(mean2) & is.na(sd2)) { return(NA) }
    if (!is.na(sd1) & is.na(sd2)) { sd2 <- sd1 }
    if (is.na(sd1) & !is.na(sd2)) { sd1 <- sd2 }
    if (sd1==0 | sd2==0) { return(NA) }
    if (is.na(mean2)) { mean2 <- 0 }

    means <- c(mean1, mean2)
    sds <- c(sd1, sd2)

    curves <- list()
    for (idx in c(1, 2)) {

        mean <- means[[idx]]
        sd <- sds[[idx]]

        if (mean==0) {
            adj_mean <- 1
        } else {
            # make the mean continuous
            signed_log <- sign(mean) * log10(1 + abs(mean))
            adj_mean <- 10^signed_log
        }
        adj_sd <- abs(sd/mean) * adj_mean

        location <- log(adj_mean^2 / sqrt(adj_sd^2 + adj_mean^2))
        shape <- sqrt(log(1 + (adj_sd / adj_mean)^2))

        curves[[idx]] <- local({
            loc <- location
            shp <- shape
            function(x) dlnorm(x, meanlog = loc, sdlog = shp)
        })
    }

    overlap <- integrate(
        function(x) pmin(curves[[1]](x), curves[[2]](x)),
        lower = 0, upper = max(mean1 + 10 * sd1, mean2 + 10 * sd2),
        rel.tol = 1e-8
    )$value

    dissimiliarity <- 1 - overlap
    return(dissimiliarity)
}

File: python/tools/math.py
import numpy as np
import pandas as pd
from scipy import stats
from itertools import combinations
from typing import List, Union
from .df_tools import pivot_then_collapse
from .list_tools import collect_matrix_cols, fill_missing_keys


def unpaired_t_test(x, y) -> float:
    """Perform unpaired t-test between two groups."""
    if x is None or y is None:
        return np.nan

    x = np.array(x).flatten() if hasattr(x, '__iter__') else np.array([x])
    y = np.array(y).flatten() if hasattr(y, '__iter__') else np.array([y])

    x = x[~np.isnan(x)]
    y = y[~np.isnan(y)]

    if len(x) == 0 or len(y) == 0:
        return np.nan
    if len(x) == 1 and len(y) == 1:
        return np.nan

    if len(x) > 1 and len(y) > 1:
        _, pval = stats.ttest_ind(x, y, equal_var=False)
        return pval

    # One-sided t-test when one group has single observation
    if len(x) == 1:
        ob, obs = x[0], y
    else:
        ob, obs = y[0], x

    alternative = 'greater' if np.mean(obs) > ob else 'less'
    _, pval = stats.ttest_1samp(obs, ob)
    return pval / 2 if alternative == 'less' else pval / 2


def apply_unpaired_t_test(df: pd.DataFrame, index_cols: List[str],
                          group_name: str, metric: str,
                          custom_group_order: List[str] = None) -> pd.DataFrame:
    """Apply unpaired t-test across a dataframe."""
    if df[metric].isna().all():
        return None

    if custom_group_order:
        group_names = [g for g in custom_group_order if g in df[group_name].unique()]
    else:
        group_names = sorted(df[group_name].unique())

    n_groups = len(group_names)
    if n_groups == 1:
        return None

    res = pivot_then_collapse(df, index_cols, group_name, metric, custom_group_order)

    # Generate pairs and compute p-values
    pairs = list(combinations(range(n_groups), 2))
    for idx1, idx2 in pairs:
        col_name = f"{group_names[idx2]}-{group_names[idx1]}"
        pvals = []
        for i in range(len(res)):
            x = res[group_names[idx1]].iloc[i]
            y = res[group_names[idx2]].iloc[i]
            pvals.append(unpaired_t_test(x, y))
        res[col_name] = pvals

    return res


def fishers_lsd(df: pd.DataFrame, group: str, metric: str,
                custom_group_order: List[str] = None) -> dict:
    """One-way ANOVA with Fisher's LSD (no correction)."""
    if df[metric].isna().all():
        return {}

    if custom_group_order:
        group_names = [g for g in custom_group_order if g in df[group].unique()]
    else:
        group_names = sorted(df[group].unique())

    n_groups = len(group_names)
    if n_groups == 1:
        return {}

    # ANOVA computations
    groups_data = [df[df[group] == g][metric].dropna().values for g in group_names]
    group_means = {g: np.mean(groups_data[i]) for i, g in enumerate(group_names)}
    group_sizes = {g: len(groups_data[i]) for i, g in enumerate(group_names)}

    overall_mean = df[metric].mean()
    ss_between = sum(group_sizes[g] * (group_means[g] - overall_mean)**2 for g in group_names)
    ss_total = ((df[metric] - overall_mean)**2).sum()
    ss_within = ss_total - ss_between

    df_within = len(df[metric].dropna()) - n_groups
    mse = ss_within / df_within

    # Fisher's LSD p-values
    pvals = {}
    for idx1, idx2 in combinations(range(n_groups), 2):
        left, right = group_names[idx1], group_names[idx2]
        col_name = f"{right}-{left}"

        mean_diff = group_means[right] - group_means[left]
        se_diff = np.sqrt(mse * (1/group_sizes[left] + 1/group_sizes[right]))
        t_stat = mean_diff / se_diff
        pval = 2 * stats.t.sf(abs(t_stat), df_within)
        pvals[col_name] = pval

    return pvals


def tukey_multiple_comparisons(df: pd.DataFrame, group: str, metric: str,
                               custom_group_order: List[str] = None) -> dict:
    """One-way ANOVA with Tukey HSD correction."""
    from scipy.stats import tukey_hsd

    if df[metric].isna().all():
        return {}

    if custom_group_order:
        group_names = [g for g in custom_group_order if g in df[group].unique()]
        df = df[df[group].isin(group_names)].copy()
    else:
        group_names = sorted(df[group].unique())

    n_groups = len(group_names)
    if n_groups == 1:
        return {}

    groups_data = [df[df[group] == g][metric].dropna().values for g in group_names]

    try:
        result = tukey_hsd(*groups_data)
        pvals = {}

        for i, j in combinations(range(n_groups), 2):
            col_name = f"{group_names[j]}-{group_names[i]}"
            pvals[col_name] = result.pvalue[i, j]

        return pvals
    except Exception:
        return fishers_lsd(df, group, metric, custom_group_order)


def bonferroni_multiple_comparisons(df: pd.DataFrame, group: str, metric: str,
                                    custom_group_order: List[str] = None) -> dict:
    """One-way ANOVA with Bonferroni correction."""
    if df[metric].isna().all():
        return {}

    if custom_group_order:
        group_names = [g for g in custom_group_order if g in df[group].unique()]
        df = df[df[group].isin(group_names)].copy()
    else:
        group_names = sorted(df[group].unique())

    n_groups = len(group_names)
    if n_groups == 1:
        return {}

    # Pairwise t-tests with Bonferroni correction
    from scipy.stats import ttest_ind

    n_comparisons = n_groups * (n_groups - 1) // 2
    pvals = {}

    for i, j in combinations(range(n_groups), 2):
        g1 = df[df[group] == group_names[i]][metric].dropna()
        g2 = df[df[group] == group_names[j]][metric].dropna()

        if len(g1) > 0 and len(g2) > 0:
            _, p = ttest_ind(g1, g2)
            col_name = f"{group_names[j]}-{group_names[i]}"
            pvals[col_name] = min(p * n_comparisons, 1.0)

    return pvals


def apply_multiple_comparisons(df: pd.DataFrame, index_cols: List[str],
                               group_name: str, metric: str,
                               correction: str = 'fishers_lsd',
                               custom_group_order: List[str] = None) -> pd.DataFrame:
    """Apply multiple comparisons across a dataframe."""
    res = pivot_then_collapse(df, index_cols, group_name, metric, custom_group_order)

    # Split dataframe for each unique combination of index columns
    grouped = df.groupby(index_cols)

    if correction == 'fishers_lsd':
        func = fishers_lsd
    elif correction == 'tukey':
        func = tukey_multiple_comparisons
    elif correction == 'bonferroni':
        func = bonferroni_multiple_comparisons
    else:
        raise ValueError("Choose correction='fishers_lsd', 'tukey', or 'bonferroni'")

    pval_list = []
    for name, group_df in grouped:
        pvals = func(group_df, group=group_name, metric=metric,
                     custom_group_order=custom_group_order)
        pval_list.append(pvals)

    # Convert to DataFrame columns
    if pval_list:
        all_cols = set()
        for p in pval_list:
            all_cols.update(p.keys())

        for col in all_cols:
            res[col] = [p.get(col, np.nan) for p in pval_list]

    return res


def generate_normal_data(n: int = 1000, mean: float = 200,
                         sd: float = 100, group_name: str = "group") -> pd.DataFrame:
    """Generate DataFrame with random normal values."""
    return pd.DataFrame({
        'group': [group_name] * n,
        'value': np.random.normal(mean, sd, n)
    })


def generate_lognormal_data(n: int = 1000, mean: float = 200,
                            sd: float = 100, group_name: str = "group") -> pd.DataFrame:
    """Generate DataFrame with lognormal values (for fluorescence simulation)."""
    if mean == 0:
        adj_mean = 1
    else:
        signed_log = np.sign(mean) * np.log10(1 + abs(mean))
        adj_mean = 10 ** signed_log

    adj_sd = abs(sd / mean) * adj_mean if mean != 0 else sd

    location = np.log(adj_mean**2 / np.sqrt(adj_sd**2 + adj_mean**2))
    shape = np.sqrt(np.log(1 + (adj_sd / adj_mean)**2))

    # Clamp location
    if location < -372.5666:
        location = -372.5666

    values = np.random.lognormal(location, shape, n)

    return pd.DataFrame({
        'group': [group_name] * n,
        'value': values
    })


def compute_normal_tvd(mean1, sd1, mean2, sd2, log_transform: bool = False) -> float:
    """Compute Total Variation Distance between two normal distributions."""
    # Handle list inputs
    if hasattr(mean1, '__iter__') and not isinstance(mean1, str):
        mean1 = np.mean(mean1)
    if hasattr(sd1, '__iter__') and not isinstance(sd1, str):
        sd1 = np.mean(sd1)
    if hasattr(mean2, '__iter__') and not isinstance(mean2, str):
        mean2 = np.mean(mean2)
    if hasattr(sd2, '__iter__') and not isinstance(sd2, str):
        sd2 = np.mean(sd2)

    if pd.isna(sd1) and pd.isna(sd2):
        return np.nan
    if pd.isna(mean2) and pd.isna(sd2):
        return np.nan
    if not pd.isna(sd1) and pd.isna(sd2):
        sd2 = sd1
    if pd.isna(sd1) and not pd.isna(sd2):
        sd1 = sd2
    if sd1 == 0 or sd2 == 0:
        return np.nan
    if pd.isna(mean2):
        mean2 = 0

    if log_transform:
        mean1 = 10**(np.sign(mean1) * np.log10(1 + abs(mean1))) / np.log(10)
        sd1 = sd1 / np.log(10)
        mean2 = 10**(np.sign(mean2) * np.log10(1 + abs(mean2))) / np.log(10)
        sd2 = sd2 / np.log(10)

    # Compute overlap using numerical integration
    from scipy.integrate import quad

    def p(x):
        return stats.norm.pdf(x, mean1, sd1)

    def q(x):
        return stats.norm.pdf(x, mean2, sd2)

    def min_pdf(x):
        return min(p(x), q(x))

    lower = min(mean1 - 10*sd1, mean2 - 10*sd2)
    upper = max(mean1 + 10*sd1, mean2 + 10*sd2)

    overlap, _ = quad(min_pdf, lower, upper)
    dissimilarity = np.sign(mean1 - mean2) * (1 - overlap)

    return dissimilarity


def compute_lognormal_tvd(mean1, sd1, mean2, sd2) -> float:
    """Compute TVD between two lognormal distributions."""
    if pd.isna(sd1) and pd.isna(sd2):
        return np.nan
    if pd.isna(mean2) and pd.isna(sd2):
        return np.nan
    if not pd.isna(sd1) and pd.isna(sd2):
        sd2 = sd1
    if pd.isna(sd1) and not pd.isna(sd2):
        sd1 = sd2
    if sd1 == 0 or sd2 == 0:
        return np.nan
    if pd.isna(mean2):
        mean2 = 0

    from scipy.integrate import quad

    params = []
    for mean, sd in [(mean1, sd1), (mean2, sd2)]:
        if mean == 0:
            adj_mean = 1
        else:
            signed_log = np.sign(mean) * np.log10(1 + abs(mean))
            adj_mean = 10 ** signed_log

        adj_sd = abs(sd / mean) * adj_mean if mean != 0 else sd
        location = np.log(adj_mean**2 / np.sqrt(adj_sd**2 + adj_mean**2))
        shape = np.sqrt(np.log(1 + (adj_sd / adj_mean)**2))
        params.append((location, shape))

    def curve1(x):
        return stats.lognorm.pdf(x, params[0][1], scale=np.exp(params[0][0]))

    def curve2(x):
        return stats.lognorm.pdf(x, params[1][1], scale=np.exp(params[1][0]))

    upper = max(mean1 + 10*sd1, mean2 + 10*sd2)
    overlap, _ = quad(lambda x: min(curve1(x), curve2(x)), 0, upper)

    dissimilarity = np.sign(mean1 - mean2) * (1 - overlap)
    return dissimilarity
